File: fresh/widgets/pQueuedMessageWidget.cpp
/****************************************************************************
	Copyright (C) 2005 - 2008  Filipe AZEVEDO & The Monkey Studio Team

	This program is free software; you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation; either version 2 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
****************************************************************************/
#include "pQueuedMessageWidget.h"

#include <QLabel>
#include <QHBoxLayout>
#include <QTimer>
#include <QShowEvent>
#include <QPushButton>
#include <QStyle>
#include <QMetaMethod>

int pQueuedMessageWidget::pQueuedMessageUniqueId = 0;

/*!
	\details Create a new pQueuedMessageWidget object
	\param parent The parent widget
*/
pQueuedMessageWidget::pQueuedMessageWidget( QWidget* parent )
	: QWidget( parent )
{
	QFont font = this->font();
	font.setPointSize( 9 );
	setFont( font );
	// pixmap
	lPixmap = new QLabel( this );
	lPixmap->setAlignment( Qt::AlignCenter );
	lPixmap->setSizePolicy( QSizePolicy( QSizePolicy::Maximum, QSizePolicy::Preferred ) );
	// message
	lMessage = new QLabel( this );
	lMessage->setAlignment( Qt::AlignVCenter | Qt::AlignLeft );
	lMessage->setSizePolicy( QSizePolicy( QSizePolicy::Expanding, QSizePolicy::Maximum ) );
	lMessage->setWordWrap( true );
	// button
	dbbButtons = new QDialogButtonBox( this );
	dbbButtons->setSizePolicy( QSizePolicy( QSizePolicy::Maximum, QSizePolicy::Preferred ) );
	// layout
	QHBoxLayout* hbl = new QHBoxLayout( this );
	hbl->setMargin( 0 );
	hbl->addWidget( lPixmap, 0, Qt::AlignCenter );
	hbl->addWidget( lMessage );
	hbl->addWidget( dbbButtons, 0, Qt::AlignCenter );
	// connections
	connect( dbbButtons, SIGNAL( clicked( QAbstractButton* ) ), this, SLOT( clicked( QAbstractButton* ) ) );
}

/*!
	\details Return the number of message currently queued
*/
int pQueuedMessageWidget::messagesCount() const
{
	return mMessages.count();
}

pQueuedMessage pQueuedMessageWidget::currentMessage() const
{
	return mMessages.values().value( 0 );
}

/*!
	\details Append a message to the queued and return it's unique \c id
	\param message The message structure to add
*/
int pQueuedMessageWidget::append( const pQueuedMessage& message )
{
	if ( !mMessages.values().contains( message ) )
	{
		mMessages[pQueuedMessageUniqueId] = message;
		if ( mMessages.count() == 1 )
			QTimer::singleShot( 0, this, SLOT( showMessage() ) );
		return pQueuedMessageUniqueId++;
	}
	return mMessages.key( message );
}

/*!
	\details Append a message to the queued and return it's unique \c id
	\param message The message to show
	\param milliseconds The milliseconds to wait before the message is auto closed, use 0 for unlimited time
	\param pixmap The pixmap to use as icon
	\param background The brush background
	\param foreground The brush foreground
*/
int pQueuedMessageWidget::append( const QString& message, int milliseconds, const QPixmap& pixmap, const QBrush& background, const QBrush& foreground )
{
	pQueuedMessage m;
	m.Message = message;
	m.MilliSeconds = milliseconds;
	m.Pixmap = pixmap;
	m.Background = background;
	m.Foreground = foreground;
	return append( m );
}

/*!
	\details Remove a message from the queue
	\param message The message to remove
*/
void pQueuedMessageWidget::remove( const pQueuedMessage& message )
{ mMessages.remove( mMessages.key( message ) ); }

/*!
	\details Remove a message from the queue by it's id
	\param id The message id to remove
*/
void pQueuedMessageWidget::remove( int id )
{ mMessages.remove( id ); }

/*!
	\details Clear the current message
*/
void pQueuedMessageWidget::clear()
{
	lPixmap->clear();
	lMessage->clear();
	dbbButtons->clear();
	emit cleared();
}

void pQueuedMessageWidget::clicked( QAbstractButton* button )
{
	const pQueuedMessage msg = mMessages.begin().value();
	if ( msg.Object && msg.Slot )
		QMetaObject::invokeMethod( msg.Object, msg.Slot, Q_ARG( QDialogButtonBox::StandardButton, dbbButtons->standardButton( button ) ), Q_ARG( pQueuedMessage, msg ) );
	closeMessage();
}

/*!
	\details Show the curernt message.
	\details The widget must be visible as only the gui contents is updated.
*/
void pQueuedMessageWidget::showMessage()
{
	// get message
	pQueuedMessage msg = mMessages.begin().value();
	// set palette
	QPalette pal = style()->standardPalette();
	if ( msg.Foreground != QBrush() )
		pal.setBrush( lMessage->foregroundRole(), msg.Foreground );
	lMessage->setPalette( pal );
	// format widget
	lPixmap->setPixmap( msg.Pixmap );
	lMessage->setText( msg.Message );
	lMessage->setToolTip( msg.Message );
	lMessage->setWhatsThis( msg.Message );
	// set buttons
	if ( msg.Buttons.isEmpty() )
		msg.Buttons[ QDialogButtonBox::Ok ] = QString();
	dbbButtons->clear();
	foreach( QDialogButtonBox::StandardButton button, msg.Buttons.keys() )
	{
		QPushButton* pb = dbbButtons->addButton( button );
		if ( !msg.Buttons[ button ].isEmpty() )
			pb->setText( msg.Buttons[ button ] );
	}
	// auto close if needed
	if ( msg.MilliSeconds > 0 )
		QTimer::singleShot( msg.MilliSeconds, this, SLOT( closeMessage() ) );
	// emit signal
	emit messageShown( msg );
}

/*!
	\details Close the current shown message
*/
void pQueuedMessageWidget::closeMessage()
{
	// emit message
	emit messageClosed( mMessages.begin().value() );
	// remove remove current message from hash
	mMessages.erase( mMessages.begin() );
	// process next if possible, else clear gui
	QTimer::singleShot( 0, this, mMessages.count() > 0 ? SLOT( showMessage() ) : SLOT( clear() ) );
	// emit finished message if needed
	if ( mMessages.count() == 0 )
		emit finished();
}

File: mks/fresh/queuedmessage/pQueuedMessageWidget.py
"""This file has been ported from fresh library by Azevedo Filippe aka PasNox

See information at https://github.com/pasnox/fresh and 
API docks at http://api.monkeystudio.org/fresh/
"""

from PyQt4.QtCore import pyqtSignal, Qt, QTimer
from PyQt4.QtGui import QAbstractButton, QColor, QBrush, QDialogButtonBox, QHBoxLayout, \
                        QLabel, QPainter, QPixmap, QSizePolicy, QWidget

class pQueuedMessage:
    def __init__(self):
        self.milliSeconds = -1
        self.background = QBrush( Qt.NoBrush )
        self.foreground = QBrush( Qt.NoBrush )
        self.slot = None
        self.buttons = {}

class pQueuedMessageWidget(QWidget):
    
    cleared = pyqtSignal()
    finished = pyqtSignal()
    shown = pyqtSignal()
    closed = pyqtSignal()
    linkActivated = pyqtSignal(unicode)
    linkHovered= pyqtSignal(unicode)
    buttonClicked = pyqtSignal(QAbstractButton)
    
    def __init__(self, *args):
        QWidget.__init__(self, *args)
        self._messages = []
        self._defaultTimeout = 0
        self._defaultPixmap = QPixmap(":/mksicons/fresh/info.png" )
        self._defaultBackground = QBrush( QColor( 250, 230, 147 ) )
        self._defaultForeground = QBrush( QColor( 0, 0, 0 ) )
        
        # pixmap
        self.lPixmap = QLabel( self )
        self.lPixmap.setAlignment( Qt.AlignCenter )
        self.lPixmap.setSizePolicy( QSizePolicy( QSizePolicy.Maximum, QSizePolicy.Preferred ) )
        
        # message
        self.lMessage = QLabel( self )
        self.lMessage.setAlignment( Qt.AlignVCenter | Qt.AlignLeft )
        self.lMessage.setSizePolicy( QSizePolicy( QSizePolicy.Expanding, QSizePolicy.Preferred ) )
        self.lMessage.setWordWrap( True )
        self.lMessage.setOpenExternalLinks( True )
        self.lMessage.setTextInteractionFlags( Qt.TextBrowserInteraction )
        
        # button
        self.dbbButtons = QDialogButtonBox( self )
        self.dbbButtons.setSizePolicy( QSizePolicy( QSizePolicy.Maximum, QSizePolicy.Preferred ) )
        
        self.setSizePolicy( QSizePolicy( QSizePolicy.Expanding, QSizePolicy.Maximum ) )
        
        # layout
        self.hbl = QHBoxLayout( self )
        self.hbl.setMargin( 0 )
        self.hbl.addWidget( self.lPixmap, 0, Qt.AlignCenter )
        self.hbl.addWidget( self.lMessage )
        self.hbl.addWidget( self.dbbButtons, 0, Qt.AlignCenter )
        
        # connections
        self.lMessage.linkActivated.connect(self.linkActivated)
        self.lMessage.linkHovered.connect(self.linkHovered)
        self.dbbButtons.clicked.connect(self.buttonClicked)

    def sizeHint(self):
        return QWidget.minimumSizeHint(self)

    def openExternalLinks(self):
        return self.lMessage.openExternalLinks()

    def defaultTimeout(self):
        return self._defaultTimeout

    def defaultPixmap(self):
        return self._defaultPixmap

    def defaultBackground(self):
        return self._defaultBackground

    def defaultForeground(self):
        return self._defaultForeground

    def currentMessageInformations(self):
        return self.currentMessagePixmap(), self.currentMessageBackground(), self.currentMessageForeground()

    def pendingMessageCount(self):
        return len(self._messages)

    def currentMessage(self):
        return self._messages[0]

    def append(self, message, milliSeconds ):
        msg = pQueuedMessage()
        msg.message = message
        if milliSeconds == -1:
            msg.milliSeconds = self._defaultTimeout
        else:
           msg.milliSeconds = milliSeconds
        msg.pixmap = self._defaultPixmap
        msg.background = self._defaultBackground
        msg.foreground = self._defaultForeground
        
        self._messages.append(msg)
            
        if  len(self._messages) == 1 :
            QTimer.singleShot( 0, self.showMessage)

    def setOpenExternalLinks(self, open ):
        self.lMessage.setOpenExternalLinks( open )

    def setDefaultTimeout(self, timeout ):
        self._defaultTimeout = timeout

    def setDefaultPixmap(self, pixmap ):
        self._defaultPixmap = pixmap

    def setDefaultBackground(self, brush ):
        self._defaultBackground = brush

    def setDefaultForeground(self, brush ):
        self._defaultForeground = brush

    def remove(self, message ):
        raise NotImplemented()  # incorrect port from cpp fresh
        if not self._messages or self._messages.first() == message:
            return

        self._messages.removeOne( message )

    def clear(self):
        self._messages.clear()
        self.lPixmap.clear()
        self.lMessage.clear()
        self.dbbButtons.clear()
        self.cleared.emit()

    def currentMessagePixmap(self):
        msg = self.currentMessage()
        if msg.pixmap.isNull():
            return self._defaultPixmap
        else:
            return msg.pixmap

    def currentMessageBackground(self):
        msg = self.currentMessage()
        if msg.background == QBrush( Qt.NoBrush ):
            return self._defaultBackground
        else:
            return msg.background

    def currentMessageForeground(self):
        msg = self.currentMessage()
        if msg.foreground == QBrush( Qt.NoBrush ):
            return self._defaultForeground
        else:
            return msg.foreground

    def paintEvent(self, event ):
        if  self.pendingMessageCount() == 0 :
            QWidget.paintEvent(self, event )
            return
        
        painter = QPainter( self )
        painter.setPen( Qt.NoPen )
        painter.setBrush( self.currentMessageBackground() )
        painter.drawRect( self.contentsRect() )

    def buttonClicked(self, button ):
        msg = self.currentMessage()
        standardButton = self.dbbButtons.standardButton( button )
        
        if msg.slot is not None:
            msg.slot(standardButton, msg)
        
        self.closeMessage()

    def showMessage(self):
        # get message
        msg = self.currentMessage()
        
        # update palette
        pal = self.lMessage.palette()
        pal.setBrush( self.lMessage.foregroundRole(), self.currentMessageForeground() )
        self.lMessage.setPalette( pal )
        
        # format widget
        self.lPixmap.setPixmap( self.currentMessagePixmap() )
        self.lMessage.setText( msg.message )
        self.lMessage.setToolTip( msg.message )
        self.lMessage.setWhatsThis( msg.message )
        
        # set buttons
        if not msg.buttons:
            msg.buttons[ QDialogButtonBox.Ok ] = None

        self.dbbButtons.clear()
        
        for button in msg.buttons.keys():
            pb = self.dbbButtons.addButton( button )
            
            if button in msg.buttons:
                pb.setText( msg.buttons[ button ] )
        
        # auto close if needed
        if msg.milliSeconds == -1:
            timeout = self._defaultTimeout
        else:
            timeout =  msg.milliSeconds
        
        if  timeout > 0:
            QTimer.singleShot( timeout, self.closeMessage )
        
        # signal.emit
        self.shown.emit()

    def closeMessage(self):
        # message.emit
        self.closed.emit()
        
        # remove remove current message from hash
        self._messages = self._messages[1:]
        
        # process next if possible, clear gui
        if self._messages:
            QTimer.singleShot( 0, self.showMessage)
        else:
            QTimer.singleShot( 0, self.clearMessage)
        
        # finished.emit message if needed
        if not self._messages:
            self.finished.emit()

    def clearMessage(self):
        self.lPixmap.clear()
        self.lMessage.clear()
        self.dbbButtons.clear()
